File: backend/ai_module/AI.py
from dotenv import load_dotenv
import os
from PIL import Image
import pytesseract
import google.generativeai as genai
from IPython.display import Markdown
import textwrap

dotenv_path = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), '.env')
load_dotenv(dotenv_path)

apiKey = os.getenv("GEMINI_API_KEY")

genai.configure(api_key=apiKey)

pytesseract.pytesseract.tesseract_cmd = r"C:\\Program Files\\Tesseract-OCR\\tesseract.exe"

def to_markdown(text):
  text = text.replace('•', '  *')
  return Markdown(textwrap.indent(text, '> ', predicate=lambda _: True))

def extract_text_from_image(image_path):
    try:
        image = Image.open(image_path)
        text = pytesseract.image_to_string(image)
        return text
    except Exception as e:
        print(f"Error extracting text from image: {e}")
        return None  # Return None on error

def generate_questions_using_gemini(prompt):
    model = genai.GenerativeModel('gemini-pro')

    try:
        response = model.generate_content(prompt)
        # print("API Request Successful")
        return to_markdown(response.text)
    except Exception as e:
        print("API Request Failed")
        print(f"Error: {e}")
        print(f"Prompt: {prompt}")  # Log the prompt for debugging
        return "Error generating question"

if __name__ == '__main__':
    # syllabus_image = sys.argv[1]

    # numberOfQuestions = sys.argv[2]

    syllabus_image = "./syll.png"
    numberOfQuestions = 15

    syllabus_text = extract_text_from_image(syllabus_image)

    prompt = f"Generate a random Multiple Choice Question paper of {numberOfQuestions} questions with 4 options A,B,C,D each based on the given syllabus:\n\n{syllabus_text} \n\n The format of the MCQs should be like this \n Each Question should start with letter 'Q.' and each option A,B,C,D should be in separate lines"

    # print(prompt)

    generated_questions = generate_questions_using_gemini(prompt)

    print(generated_questions)

File: backend/ai_module/AI.js
const { GoogleGenerativeAI } = require("@google/generative-ai");
const Tesseract = require('tesseract.js');
require('dotenv').config();

async function extractTextFromImage(imagePath) {
    try {
        const text = await Tesseract.recognize(imagePath, 'eng'); 
        return text.data.text;
    } catch (error) {
        console.error("Error extracting text from image:", error);
        return null;
    }
}
const apiKey=process.env.GEMINI_API_KEY;
// console.log(apiKey)

async function generateRandomQuestions(syllabusPath, numberOfQuestions) {
    const syllabusText = await extractTextFromImage(syllabusPath);
    const sampleQuestion = "**Q. A sample Question#\n**A. option A#\n**B. option B#\n**C. option C#\n**D. option D#\n**Ans. A/B/C/D#";

    const genAI = new GoogleGenerativeAI(apiKey);
    
    const model = genAI.getGenerativeModel({ model: "gemini-pro"});

    const prompt = `Generate a random Multiple Choice Question paper of ${numberOfQuestions} questions with 4 options A,B,C,D each based on the given syllabus:\n\n${syllabusText} \n\n The format of the MCQs should be like this \n Each Question should start with letter '**Q.' and each option A,B,C,D should be in separate lines,and at the last line should be the correct option i.e the answer of the questions, also '#' should be used to mark the end of question and each option like this : \n ${sampleQuestion}`;
    
    try {
      const result = await model.generateContent(prompt);
      const response = result.response;
      return response.text();
    } catch (error) {
        console.error("Error generating questions:", error);
    }
  }
// generateRandomQuestions()
module.exports = generateRandomQuestions;